index.js

```diff
@@ -124,6 +151,298 @@
     for (var n = start; n < stop; n += 1) result.push(n);
     return result;
   };
+  
+  //  values :: Object -> [Any]
+  var values = function(obj) {
+      var result = [];
+      for (var key in obj) if (has(key, obj)) result.push(obj[key]);
+      return result;
+  };
+  
+  var strRepeat = function(s, times) {
+    return Array(times + 1).join(s);
+  };
+  
+  var findVar = function(name) {
+    return function(acc, token) {
+      return (
+        isVar(token) && token.text === name ? acc.concat(token.start) :
+        acc
+      );
+    };
+  };
+  
+  var offsets = function(sig, name) {
+    var idx, pos;
+    var positions = traverse(findVar(name), [], sig);
+    var upcarrots = '';
+    var varNumbers = '';
+    for (idx = 0; idx < positions.length; idx += 1) {
+      pos = positions[idx];
+      upcarrots += strRepeat(' ', pos - upcarrots.length) + '^';
+      varNumbers += strRepeat(' ', pos - varNumbers.length) + (idx + 1);
+    }
+    return upcarrots + '\n' + varNumbers;
+  };
+  
+  // closure.. no reason really beyond helping with name clashes
+  var signature = (function(){
+    
+    var Grammar = function(open, close, prefix, glue, separator) {
```

Comment:
As I find is often the case with configuration, this seemed like a good idea at first but I'm not sure there is much use it. Pretty much everything is a special case, and the functions would be so simple anyway I don't know that this is the most useful way to define things.